firedrake/assemble.py

```diff
@@ -381,167 +399,164 @@ def vec(testmap, i):
-            # end (facet number information).
-            extra_args = []
-            # Decoration for applying to matrix maps in extruded case
-            decoration = None
-            itspace = m.measure_set(integral_type, subdomain_id,
-                                    all_integer_subdomain_ids)
-            if integral_type == "cell":
-                itspace = sdata or itspace
-
-                if subdomain_id not in ["otherwise", "everywhere"] and \
-                   sdata is not None:
-                    raise ValueError("Cannot use subdomain data and subdomain_id")
-
-                def get_map(x):
-                    return x.cell_node_map()
-
-            elif integral_type in ("exterior_facet", "exterior_facet_vert"):
-                extra_args.append(m.exterior_facets.local_facet_dat(op2.READ))
-
-                def get_map(x):
-                    return x.exterior_facet_node_map()
-
-            elif integral_type in ("exterior_facet_top", "exterior_facet_bottom"):
-                # In the case of extruded meshes with horizontal facet integrals, two
-                # parallel loops will (potentially) get created and called based on the
-                # domain id: interior horizontal, bottom or top.
-                decoration = {"exterior_facet_top": op2.ON_TOP,
-                              "exterior_facet_bottom": op2.ON_BOTTOM}[integral_type]
-                kwargs["iterate"] = decoration
-
-                def get_map(x):
-                    return x.cell_node_map()
-
-            elif integral_type in ("interior_facet", "interior_facet_vert"):
-                extra_args.append(m.interior_facets.local_facet_dat(op2.READ))
-
-                def get_map(x):
-                    return x.interior_facet_node_map()
-
-            elif integral_type == "interior_facet_horiz":
-                decoration = op2.ON_INTERIOR_FACETS
-                kwargs["iterate"] = decoration
-
-                def get_map(x):
-                    return x.cell_node_map()
+                        if isinstance(bc, DirichletBC):
+                            trbc.append(bc)
+                        elif isinstance(bc, EquationBCSplit):
+                            collect_dbc(bc.bcs)
+            collect_dbc(bcs)
+
+        # Now build arguments for the par_loop
+        kwargs = {}
+        # Some integrals require non-coefficient arguments at the
+        # end (facet number information).
+        extra_args = []
+        # Decoration for applying to matrix maps in extruded case
+        decoration = None
+        itspace = m.measure_set(integral_type, subdomain_id,
+                                all_integer_subdomain_ids)
+        if integral_type == "cell":
+            itspace = sdata or itspace
+            if subdomain_id not in ["otherwise", "everywhere"] and sdata is not None:
+                raise ValueError("Cannot use subdomain data and subdomain_id")
+
+            def get_map(x):
```

Comment:
I think `get_map` can also be parameterized by the arguments `x` and `integral_type`.

Reply:
It's defined differently based on the integral_type right now.